// yandex-dialogs-sdk/dist/reply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint camelcase: 0 */
const constants_1 = require("./constants");
function reply(params) {
    const data = {
        response: {
            text: '',
            buttons: [],
            end_session: constants_1.DEFAULT_END_SESSION,
        },
        session: null,
        version: constants_1.ALICE_PROTOCOL_VERSION,
    };
    if (typeof params === 'string') {
        data.response.text = params;
        return data;
    }
    else if (typeof params === 'object') {
        const { text, tts, endSession, end_session, session, buttons, card } = params;
        if (text) {
            data.response.text = text;
        }
        if (tts) {
            data.response.tts = tts;
        }
        if (buttons) {
            data.response.buttons = buttons;
        }
        if (card) {
            data.response.card = card;
        }
        if (end_session || endSession) {
            data.response.end_session = end_session || endSession;
        }
        if (session) {
            data.session = session;
        }
        return data;
    }
    else {
        throw new Error('reply: invalid parameter. Use Object or String instead');
    }
}
exports.reply = reply;
exports.default = reply;
